Added pronunciation dictionary tutorial (#216)
elevenlabs/elevenlabs-docs: api-reference/how-to-use-pronunciation-dictionary-in-python.mdx

---
title: "How to use pronunciation dictionary (Python SDK)"
sidebarTitle: "How to use pronunciation dictionary"
description: "How to add, view, and remove rules to pronunciation dictionary with Python SDK"
icon: "code"
---

> In this tutorial, you'll learn how to use a pronunciation dictionary with the ElevenLabs Python SDK. Pronunciation dictionaries are useful for controlling the specific pronunciation of words. We support both [IPA](https://en.wikipedia.org/wiki/International_Phonetic_Alphabet) or [CMU](https://en.wikipedia.org/wiki/CMU_Pronouncing_Dictionary) alphabets. It is useful for correcting rare or specific pronunciations, such as names or companies. For example, the word `nginx` could be pronounced incorrectly. Instead, we can add our version of pronunciation. Based on IPA, `nginx` is pronounced as `/ˈɛndʒɪnˈɛks/`. Finding IPA or CMU of words manually can be difficult. Instead, LLMs like ChatGPT can help you to make the search easier.

Reviewer: ```suggestion
In this tutorial, you'll learn how to use a pronunciation dictionary with the ElevenLabs Python SDK. Pronunciation dictionaries are useful for controlling the specific pronunciation of words. We support both [IPA](https://en.wikipedia.org/wiki/International_Phonetic_Alphabet) and [CMU](https://en.wikipedia.org/wiki/CMU_Pronouncing_Dictionary) alphabets. It is useful for correcting rare or specific pronunciations, such as names or companies. For example, the word `nginx` could be pronounced incorrectly. Instead, we can add our version of pronunciation. Based on IPA, `nginx` is pronounced as `/ˈɛndʒɪnˈɛks/`. Finding IPA or CMU of words manually can be difficult. Instead, LLMs like ChatGPT can help you to make the search easier.

```

We'll start by adding rules to the pronunciation dictionary from a file and comparing the text-to-speech results that use and do not use the dictionary. After that, we'll discuss how to add and remove specific rules to existing dictionaries.

> If you want to jump straight to the finished repo you can find it [here](https://github.com/elevenlabs/elevenlabs-examples/tree/main/pronunciation-dictionaries/examples/pronunciation-dictionaries/python)

Reviewer: ```suggestion
If you want to jump straight to the finished repo you can find it [here](https://github.com/elevenlabs/elevenlabs-examples/tree/main/examples/pronunciation-dictionaries/python)

```

            type="phoneme",
            alphabet="ipa",
            string_to_replace="tomato",
            phoneme="/tə'meɪtoʊ/",
        ),
        BodyAddRulesToThePronunciationDictionaryV1PronunciationDictionariesPronunciationDictionaryIdAddRulesPostRulesItem_Phoneme(
            type="phoneme",
            alphabet="ipa",
            string_to_replace="Tomato",
            phoneme="/tə'meɪtoʊ/",
        ),
    ],
)

audio_4 = client.generate(
    text="With the rule added again: tomato",
    voice="Rachel",
    model="eleven_turbo_v2",
    pronunciation_dictionary_locators=[
        PronunciationDictionaryVersionLocator(
            pronunciation_dictionary_id=pronunciation_dictionary_rules_added.id,
            version_id=pronunciation_dictionary_rules_added.version_id,
        )
    ],
)

play(audio_4)
```

## Conclusion

You know how to use a pronunciation dictionary for generating text-to-speech audio. These functionailities open up opportunities to generate text-to-speech audio based on your pronunciation dictionary, making it more flexible for your use case.

> For more details, visit our [example repo](https://github.com/elevenlabs/elevenlabs-examples/tree/main/pronunciation-dictionaries/examples/pronunciation-dictionaries/python) to see the full project files which give a clear structure for setting up your application:

Reviewer: ```suggestion
For more details, visit our [example repo](https://github.com/elevenlabs/elevenlabs-examples/tree/main/examples/pronunciation-dictionaries/python) to see the full project files which give a clear structure for setting up your application:

```

---
title: "How to use pronunciation dictionary (Python SDK)"
sidebarTitle: "How to use pronunciation dictionary"

> description: "How to add, view, and remove rules to pronunciation dictionary with Python SDK"

Reviewer: ```suggestion
description: "How to add, view, and remove rules to pronunciation dictionaries with the Python SDK"

```

---

> title: "How to use pronunciation dictionary (Python SDK)"
> sidebarTitle: "How to use pronunciation dictionary"

Reviewer: ```suggestion
title: "How to use pronunciation dictionaries (Python SDK)"
sidebarTitle: "How to use pronunciation dictionaries"

```

---
title: "How to use pronunciation dictionary (Python SDK)"
sidebarTitle: "How to use pronunciation dictionary"
description: "How to add, view, and remove rules to pronunciation dictionary with Python SDK"
icon: "code"
---

> In this tutorial, you'll learn how to use a pronunciation dictionary with the ElevenLabs Python SDK. We'll start by adding rules to the pronunciation dictionary from a file and then compare the text-to-speech results that use and do not use the dictionary. After that, we'll remove the rule and test it again. Finally, we'll add a rule from the dictionary rules and test it once more.

Reviewer: ```suggestion
In this tutorial, you'll learn how to use a pronunciation dictionary with the ElevenLabs Python SDK. Pronunciation dictionaries are useful for controlling the specific pronunciation of words. We support both [IPA](https://en.wikipedia.org/wiki/International_Phonetic_Alphabet) or [CMU](https://en.wikipedia.org/wiki/CMU_Pronouncing_Dictionary) alphabets.

We'll start by adding rules to the pronunciation dictionary from a file and comparing the text-to-speech results that use and do not use the dictionary. After that, we'll discuss how to add and remove specific rules to existing dictionaries.
```

Reviewer: Also include the "Why" - add a few examples. Like controlling the pronunciation of names and companies

Reviewer: somewhere can you include that chatGPT is good for finding the IPA or CMU of words

Author: Done. Please check the latest commit

> If you want to jump straight to the finished repo you can find it [here](https://github.com/elevenlabs/elevenlabs-examples/tree/main/pronunciation-dictionaries/examples/pronunciation-dictionaries/python)

Reviewer: Luke to check link after merging

## Requirements

- An ElevenLabs account with an API key (here’s how to [find your API key](/api-reference/text-to-speech#authentication)).
- Python installed on your machine
- FFMPEG to play audio

## Setup

### Installing our SDK

Before you begin, make sure you have installed the necessary SDKs and libraries. You will need the ElevenLabs SDK for the updating pronunciation dictionary and using text-to-speech conversion. You can install it using pip:

```bash
pip install elevenlabs
```

Additionally, install `python-dotenv` to manage your environmental variables:

```bash
pip install python-dotenv
```

Next, create a `.env` file in your project directory and fill it with your credentials like so:

```
ELEVENLABS_API_KEY=your_elevenlabs_api_key_here
```

## Initiate the Client SDK

We'll start by initializing the client SDK.

```python
import os
from elevenlabs.client import ElevenLabs

ELEVENLABS_API_KEY = os.getenv("ELEVENLABS_API_KEY")
client = ElevenLabs(
    api_key=ELEVENLABS_API_KEY,
)
```

## Create a Pronunciation Dictionary From a File

> To create a pronunciation dictionary from a File, we'll start by creating a `.pls` file for our rules.

Reviewer: ```suggestion
To create a pronunciation dictionary from a File, we'll create a `.pls` file for our rules.

This rule will use the "IPA" alphabet and update the pronunciation for `tomato` and `Tomato` with a different pronunciation. PLS files are case sensitive which is why we include it both with and without a capital "T". Save it as `dictionary.pls`.
```

Author: Done

> ```xml

Reviewer: add name of the file

Author: Done

<?xml version="1.0" encoding="UTF-8"?>
<lexicon version="1.0" 
      xmlns="http://www.w3.org/2005/01/pronunciation-lexicon"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
      xsi:schemaLocation="http://www.w3.org/2005/01/pronunciation-lexicon 
        http://www.w3.org/TR/2007/CR-pronunciation-lexicon-20071212/pls.xsd"
      alphabet="ipa" xml:lang="en-US">
  <lexeme>
    <grapheme>tomato</grapheme>
    <phoneme>/tə'meɪtoʊ/</phoneme>
  </lexeme>
  <lexeme>
    <grapheme>Tomato</grapheme>
    <phoneme>/tə'meɪtoʊ/</phoneme>
  </lexeme>
</lexicon>
```

> This rule will use "ipa" alphabet and update pronunciation for `tomato` and `Tomato` with phoneme `/tə'meɪtoʊ/`. Save it as `dictionary.pls`.

Reviewer: ```suggestion
```

```python
import requests
from elevenlabs import play, PronunciationDictionaryVersionLocator

> def print_rules(client: ElevenLabs, dictionary_id: str, version_id: str):
>     # TODO: replace with the SDK when fixed
>     # client.pronunciation_dictionary.get_pls_file_with_a_pronunciation_dictionary_version_rules(
>     #     dictionary_id=dictionary_id,
>     #     version_id=version_id,
>     # )
>     response = requests.get(
>         f"https://api.elevenlabs.io/v1/pronunciation-dictionaries/{dictionary_id}/{version_id}/download",
>         headers={"xi-api-key": ELEVENLABS_API_KEY},
>     )
>     print("rules", response.text)

Reviewer: ```suggestion
```

with open("dictionary.pls", "rb") as f:
    # this dictionary changes how tomato is pronounced
    pronunciation_dictionary = client.pronunciation_dictionary.add_from_file(
        file=f.read(), name="example"
    )

> print("-- initial rules --")

Reviewer: ```suggestion
```

> print_rules(
>     client, pronunciation_dictionary.id, pronunciation_dictionary.version_id
> )

Reviewer: ```suggestion
```
This is useful for the main code block but not needed for the tutorial

audio_1 = client.generate(
    text="Without the dictionary: tomato",
    voice="Rachel",
    model="eleven_turbo_v2",
)

audio_2 = client.generate(
    text="With the dictionary: tomato",
    voice="Rachel",
    model="eleven_turbo_v2",
    pronunciation_dictionary_locators=[
        PronunciationDictionaryVersionLocator(
            pronunciation_dictionary_id=pronunciation_dictionary.id,
            version_id=pronunciation_dictionary.version_id,
        )
    ],
)

# play the audio
play(audio_1)
play(audio_2)
```

> In the snippet, we start by adding rules from a file and get the uploaded result. Finally, we generate and play two different text-to-speech audio to compare the custom pronunciation dictionary.

Author: @lharries I'm debating myself whether to add this type of brief description every subsection. I am leaning towards to remove but I keep it here in case you disagree with me. 

The purpose of this brief two sentences to explain on what the snippet does in general but I think the code should self-explanatory by itself. If you disagree, please feel free to leave it as is

Reviewer: I think it's useful if it adds some extra context, but I'd add it above the code block

## Remove Rules From a Pronunciation Dictionary

> To remove rules from a pronunciation dictionary, we can simply call `remove_rules_from_the_pronunciation_dictionary` method in the pronunciation dictionary module.

Reviewer: Mention that every time you edit the pronunciation dictionary, that it returns a pronunciation dictionary with a new version ID - and that you need to use this. The old dictionary will also persist

```python
pronunciation_dictionary_rules_removed = (
    client.pronunciation_dictionary.remove_rules_from_the_pronunciation_dictionary(
        pronunciation_dictionary_id=pronunciation_dictionary.id,
        rule_strings=["tomato", "Tomato"],
    )
)

> print("\n\n-- removed rule --\n\n")
> 
> print_rules(
>     client,
>     pronunciation_dictionary_rules_removed.id,
>     pronunciation_dictionary_rules_removed.version_id,
> )

Reviewer: ```suggestion
```

audio_3 = client.generate(
    text="With the rule removed: tomato",
    voice="Rachel",
    model="eleven_turbo_v2",
    pronunciation_dictionary_locators=[
        PronunciationDictionaryVersionLocator(
            pronunciation_dictionary_id=pronunciation_dictionary_rules_removed.id,
            version_id=pronunciation_dictionary_rules_removed.version_id,
        )
    ],
)

play(audio_3)
```

> In the snippet, we start by removing rules based on the rule string and get the updated result. Finally, we generate and play another text-to-speech audio to test the difference.

Reviewer: ```suggestion
```
This one doesn't add any more than the block below, so we should remove it

```python
from elevenlabs import BodyAddRulesToThePronunciationDictionaryV1PronunciationDictionariesPronunciationDictionaryIdAddRulesPostRulesItem_Phoneme

pronunciation_dictionary_rules_added = client.pronunciation_dictionary.add_rules_to_the_pronunciation_dictionary(
    pronunciation_dictionary_id=pronunciation_dictionary_rules_removed.id,
    rules=[
        # TODO rename this
        BodyAddRulesToThePronunciationDictionaryV1PronunciationDictionariesPronunciationDictionaryIdAddRulesPostRulesItem_Phoneme(
            type="phoneme",
            alphabet="ipa",
            string_to_replace="tomato",
            phoneme="/tə'meɪtoʊ/",
        ),
        BodyAddRulesToThePronunciationDictionaryV1PronunciationDictionariesPronunciationDictionaryIdAddRulesPostRulesItem_Phoneme(
            type="phoneme",
            alphabet="ipa",
            string_to_replace="Tomato",
            phoneme="/tə'meɪtoʊ/",
        ),
    ],
)

> print("\n\n-- added rule --\n\n")
> 
> print_rules(
>     client,
>     pronunciation_dictionary_rules_added.id,
>     pronunciation_dictionary_rules_added.version_id,
> )

Reviewer: ```suggestion
```

audio_4 = client.generate(
    text="With the rule added again: tomato",
    voice="Rachel",
    model="eleven_turbo_v2",
    pronunciation_dictionary_locators=[
        PronunciationDictionaryVersionLocator(
            pronunciation_dictionary_id=pronunciation_dictionary_rules_added.id,
            version_id=pronunciation_dictionary_rules_added.version_id,
        )
    ],
)

play(audio_4)
```

> In the snippet, we start by adding rules with `BodyAddRulesToThePronunciationDictionaryV1PronunciationDictionariesPronunciationDictionaryIdAddRulesPostRulesItem_Phoneme` class and get the updated XPathResult. Finally, we generate and play another text-to-speech audio to test the difference.

Reviewer: ```suggestion
```
This snippet doesn't add any more than the line above

## Conclusion

> You know how to use a pronunciation dictionary for generating text-to-speech audio. These functionailities open up opportunities to generate text-to-speech audio based on your pronunciation dictionary, making it more flexible for your use case.

Reviewer: include some links to the other tutorials for SEO purposes. also say some suggestions of what they could build with it
